### PR TITLE
fix traceback when interrupting run

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?
Adds signal handler for user-initiated termination (SIGINT / SIGTERM) to avoid long weird traceback. Related issue: https://github.com/NVIDIA/Megatron-LM/issues/3260

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by janEbert** [comment]
I remember the TorchTitan team also having a few goes at this until they settled on the following:
https://github.com/pytorch/torchtitan/blob/10d8a306a035988a83c54c7870a41f1813f57825/torchtitan/train.py#L748-L777

Maybe it's helpful to take a look at it. Maybe it's not exactly related to this, since it was more about a proper process group shutdown.

File: megatron/core/__init__.py [hunk 1]
```diff
@@ -50,3 +55,35 @@
...
+    """
+    Signal handler for user-initiated termination (SIGINT / SIGTERM).
+
+    This handler attempts a best-effort graceful shutdown:
+      - Logs a single termination message from rank 0
+      - Synchronizes all ranks (barrier)
+      - Destroys the distributed process group
+      - Exits the process cleanly
+    """
+    from megatron.training import print_rank_0
```
Comment (jaredcasper) [inline megatron/core/__init__.py]: I don't think we want to use megatron.training inside megatron.core, or print anything for that matter.

File: megatron/core/__init__.py [hunk 2]
```diff
@@ -50,3 +55,35 @@
...
+
+    print_rank_0("\nTermination requested. Performing orderly shutdown.")
+
+    try:
+        if dist.is_available() and dist.is_initialized():
+            # synchronize all ranks before exiting
+            dist.barrier()
```
Comment (jaredcasper) [inline megatron/core/__init__.py]: What if the signal isn't sent to all ranks?

File: megatron/core/__init__.py [hunk 3]
```diff
@@ -50,3 +55,35 @@
...
+    except Exception:
+        pass
+
+    sys.exit(0)
+
+
+# Register signal handlers for both:
+#  - SIGINT  (Ctrl+C from user)
+#  - SIGTERM (sent by torchrun to worker processes)
+signal.signal(signal.SIGINT, graceful_shutdown)
```
Comment (jaredcasper) [inline megatron/core/__init__.py]: Does this remove all signal handling? What if the front end wants to do its own signal handling?

File: megatron/core/__init__.py (hunk 2 above)
Comment (dimapihtar) [reply megatron/core/__init__.py]: fixed

File: megatron/core/__init__.py (hunk 1 above)
Comment (dimapihtar) [reply megatron/core/__init__.py]: moved to megatron/training

File: megatron/core/__init__.py (hunk 3 above)
Comment (dimapihtar) [reply megatron/core/__init__.py]: also fixed
added `exit_signal_handler_for_training` flag so user can disable/enable it if needed.
